File: py/packages/sdk/src/antfly/index_maintenance.py
"""Bounded validation for exact-proof shared index maintenance operations."""

import re
from typing import Any

from .client_generated.models import IndexMaintenanceOwnerProof, IndexMaintenanceRequest, IndexMaintenanceResponse
from .exceptions import AntflyException


def _decimal(value: Any, positive: bool) -> bool:
    return (
        isinstance(value, str)
        and re.fullmatch(r"0|[1-9][0-9]{0,19}", value) is not None
        and int(value) <= 2**64 - 1
        and (not positive or value != "0")
    )


def _digest(value: Any) -> bool:
    return isinstance(value, str) and re.fullmatch(r"[0-9a-f]{64}", value) is not None


def validate_request(request: IndexMaintenanceRequest) -> set[str]:
    if (
        not isinstance(request, IndexMaintenanceRequest)
        or not _decimal(request.table_id, True)
        or type(request.schema_version) is not int
        or not 0 <= request.schema_version <= 2**32 - 1
        or not isinstance(request.owners, list)
        or not 1 <= len(request.owners) <= 128
    ):
        raise ValueError("index maintenance requires a valid table/schema identity and 1..128 owner proofs")
    groups: set[str] = set()
    for owner in request.owners:
        if (
            not isinstance(owner, IndexMaintenanceOwnerProof)
            or not _decimal(owner.group_id, True)
            or not _decimal(owner.generation, True)
            or not _decimal(owner.maintenance_epoch, False)
            or type(owner.slot) is not int
            or not 0 <= owner.slot <= 2**32 - 1
            or not _digest(owner.owner)
            or not _digest(owner.comparison)
            or not _digest(owner.progress_digest)
            or owner.group_id in groups
        ):
            raise ValueError("index maintenance has an invalid or duplicate owner proof")
        groups.add(owner.group_id)
    return groups


def validate_response(value: Any, groups: set[str]) -> IndexMaintenanceResponse:
    if not isinstance(value, dict) or not isinstance(value.get("acknowledged_groups"), list):
        raise AntflyException("invalid index maintenance acknowledgement; resubmit identical proofs")
    acknowledged = value["acknowledged_groups"]
    if (
        len(acknowledged) != len(groups)
        or any(not isinstance(group, str) for group in acknowledged)
        or len(set(acknowledged)) != len(acknowledged)
        or set(acknowledged) != groups
    ):
        raise AntflyException("invalid or incomplete index maintenance acknowledgement set; resubmit identical proofs")
    return IndexMaintenanceResponse.from_dict(value)

File: ts/packages/sdk/src/index-maintenance.ts
import type { IndexMaintenanceRequest, IndexMaintenanceResponse } from "./types.js";

function decimal(value: unknown, positive: boolean): boolean {
  return (
    typeof value === "string" &&
    /^(0|[1-9][0-9]{0,19})$/.test(value) &&
    BigInt(value) <= 18446744073709551615n &&
    (!positive || value !== "0")
  );
}

function digest(value: unknown): boolean {
  return typeof value === "string" && /^[0-9a-f]{64}$/.test(value);
}

export function validateIndexMaintenanceRequest(request: IndexMaintenanceRequest): Set<string> {
  if (
    !request ||
    !decimal(request.table_id, true) ||
    !Number.isInteger(request.schema_version) ||
    request.schema_version < 0 ||
    request.schema_version > 4294967295 ||
    !Array.isArray(request.owners) ||
    request.owners.length < 1 ||
    request.owners.length > 128
  ) {
    throw new Error(
      "Index maintenance requires a valid table/schema identity and 1..128 owner proofs"
    );
  }
  const groups = new Set<string>();
  for (const owner of request.owners) {
    if (
      !owner ||
      !decimal(owner.group_id, true) ||
      !decimal(owner.generation, true) ||
      !decimal(owner.maintenance_epoch, false) ||
      !Number.isInteger(owner.slot) ||
      owner.slot < 0 ||
      owner.slot > 4294967295 ||
      !digest(owner.owner) ||
      !digest(owner.comparison) ||
      !digest(owner.progress_digest) ||
      groups.has(owner.group_id)
    ) {
      throw new Error("Index maintenance has an invalid or duplicate owner proof");
    }
    groups.add(owner.group_id);
  }
  return groups;
}

export function validateIndexMaintenanceResponse(
  value: unknown,
  groups: Set<string>
): IndexMaintenanceResponse {
  const response = value as IndexMaintenanceResponse | null;
  if (
    !response ||
    !Array.isArray(response.acknowledged_groups) ||
    response.acknowledged_groups.length !== groups.size
  ) {
    throw new Error("Incomplete index maintenance acknowledgement; resubmit identical proofs");
  }
  const remaining = new Set(groups);
  for (const group of response.acknowledged_groups) {
    if (typeof group !== "string" || !remaining.delete(group)) {
      throw new Error("Invalid index maintenance acknowledgement set; resubmit identical proofs");
    }
  }
  return response;
}
